### ISSUE TITLE
Pyright finds a "constant" in code that does not have a constant.

### ISSUE THREAD

**Issue by hunterhogan** [microsoft/pyright#10331]
# Describe the bug

Pyright finds a "constant" in code that does not have a constant.

## Code or Screenshots

```python
I_AM_A_PYTHON_VARIABLE = 'Please, do not change me.'

I_AM_A_PYTHON_VARIABLE = 'The sky is falling.'
```

```json
[{
    "resource": "/c:/sacrilege/violations.py",
    "owner": "python",
    "code": {
        "value": "reportConstantRedefinition",
        "target": {
            "$mid": 1,
            "path": "/microsoft/pyright/blob/main/docs/configuration.md",
            "scheme": "https",
            "authority": "github.com",
            "fragment": "reportConstantRedefinition"
        }
    },
    "severity": 8,
    "message": "'I_AM_A_PYTHON_VARIABLE' is constant (because it is uppercase) and cannot be redefined",
    "source": "Pylance",
    "startLineNumber": 666,
    "startColumn": 1,
    "endLineNumber": 666,
    "endColumn": eternity
}]

```

## VS Code extension or command-line

> Version: 1.100.0-insider (user setup)
> Commit: c73f02083edc88e5c72a8c6cc301fb8a6430493f
> Date: 2025-04-15T05:04:09.439Z
> Electron: 34.4.1
> ElectronBuildId: 11317338
> Chromium: 132.0.6834.210
> Node.js: 20.18.3
> V8: 13.2.152.41-electron.0
> OS: Windows_NT x64 10.0.26100
>
> ms-python.vscode-pylance
> Version 2025.4.1

<https://github.com/microsoft/pyright/blob/41b8d83170e80acb2999a07f1341505ad6221c7a/docs/configuration.md?plain=1#L160>

> * reportConstantRedefinition [boolean or string, optional]: Generate or suppress diagnostics for attempts to redefine variables whose names are all-caps with underscores and numerals. The default value for this setting is "none".

## Solution

1. Change the patently wrong "reportConstantRedefinition" to "reportAllCapsVariableNameRedefinition".
2. Move the diagnostic rule to the newly created type checking mode: "pydogma".

| Diagnostic Rule                       | Off       | Basic   | Standard | Strict  | PyDogma |
| :------------------------------------ | :-------- | :------ | :------- | :------ | :------ |
| enableReachabilityAnalysis            | false     | true    | true     | true    | true    |
| strictListInference                   | false     | false   | false    | true    | true    |
| reportMissingTypeStubs                | "none"    | "none"  | "none"   | "error" | "error" |
| reportAllCapsVariableNameRedefinition | false     | false   | false    | false   | "sin"   |
| reportInvalidTypeForm                 | "warning" | "error" | "error"  | "error" | "error" |

## Summary

I cannot _redefine_ a Python constant because I cannot _define_ a Python constant, therefore, a diagnostic rule chastizing me for something I did not do (because it is impossible to do) is nonsensical. Keep the rule: name it properly and categorize it properly.

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

Most programming languages have a keyword such as `const` to declare that a symbol should be treated as constant. Python has no such keyword, so conventions are used instead. By convention, symbols that are named using all-caps are intended to be constants. Pyright is honoring this convention.

**Comment by hunterhogan** [comment]
"as designed" != "as aught to be designed"
